File: 2017/js/day4.js
window.AdventOfCode.Day4 = function( input )
{
	input = input.split( '\n' ).map( ( pass ) => pass.split( ' ' ) );

	const countValid = ( counter, pass ) =>
	{
		if( ( new Set( pass ) ).size === pass.length )
		{
			counter++;
		}

		return counter;
	};

	const part1 = input.reduce( countValid, 0 );

	const part2 = input
		.map( ( pass ) => pass.map( ( word ) => [ ...word ].sort( ).join( '' ) ) )
		.reduce( countValid, 0 );

	return [ part1, part2 ];
};

File: 2022/Answers/Solutions/Day4.cs
namespace AdventOfCode;

[Answer(4)]
public class Day4 : IAnswer
{
	public (string Part1, string Part2) Solve(string input)
	{
		var i = 0;
		var length = input.Length;
		var part1 = 0;
		var part2 = 0;

		int ParseIntUntil(char c)
		{
			var result = 0;

			do
			{
				var t = input[i++];

				if (t == c)
				{
					break;
				}

				result = 10 * result + t - '0';
			}
			while (i < length);

			return result;
		}

		while (i < length)
		{
			var a = ParseIntUntil('-');
			var b = ParseIntUntil(',');
			var c = ParseIntUntil('-');
			var d = ParseIntUntil('\n');

			if ((a >= c && b <= d) || (c >= a && d <= b))
			{
				part1++;
			}

			if (b >= c && a <= d)
			{
				part2++;
			}
		}

		return (part1.ToString(), part2.ToString());
	}
}
